_posts/2018-09-04-http-part4.markdown

```diff
@@ -0,0 +1,236 @@
+
+``` kotlin
+class Server(
+        val handler: Handler,
+        val port: Int,
+        val maxIncomingConnections: Int = 10,
+        val maxWorkerThreads: Int = 50
+): AutoCloseable {
+
+    private val logger = LoggerFactory.getLogger(Server::class.java)
+
+    private val serverSocket = ServerSocket(port, maxIncomingConnections)
+    private val handlerThreadPool = Executors.newFixedThreadPool(maxWorkerThreads) // 1
+    private val acceptingConnectionsThreadPool = Executors.newSingleThreadExecutor()
+
+    init {
+        logger.info("Starting server on port $port")
+        acceptingConnectionsThreadPool.submit(this::acceptConnections)
+    }
+
+    private fun acceptConnections() {
+        while (!serverSocket.isClosed) {
+            try {
+                acceptConnection()
+            } catch (e: Exception) {
+                when {
+                    e is InterruptedException -> throw e
+                    e is SocketException && serverSocket.isClosed -> logger.trace("Socket was closed", e)
+                    else -> logger.error("Could not accept new connection", e)
+                }
+            }
+        }
+    }
+
+    private fun acceptConnection() {
+        val socket = serverSocket.accept() // 2
+        handlerThreadPool.submit { // 3
+            try {
+                handleConnection(socket)
+            } catch (e: Exception) {
+                logger.warn("Could not handle connection", e)
+            }
+        }
+    }
+
+    private fun handleConnection(socket: Socket) {
+        socket.use {
+            val input = it.getInputStream().bufferedReader()
+            val output = PrintWriter(it.getOutputStream())
+
+            val response = try {
+                val request = Request.fromRaw(input)
+                handler(request)
+            } catch (e: RequestParseException) {
+                logger.warn("Could not parse a request", e)
+                Response(status = Status.BAD_REQUEST)
+            }
+            output.print(response.toRaw())
+            output.flush()
+        }
+    }
+
+    override fun close() {
+        logger.info("Stopping server")
+        serverSocket.close()
+        handlerThreadPool.shutdown()
+        handlerThreadPool.awaitTermination(1, TimeUnit.SECONDS)
+        acceptingConnectionsThreadPool.shutdown()
+        acceptingConnectionsThreadPool.awaitTermination(1, TimeUnit.SECONDS)
+    }
+}
+```
+
+We added a thread pool (1) of max worker threads of 20. Now our server blocks on the `acceptingConnectionsThreadPool` to listen for a connection (2). When we've got the connection we immediately delegate reading a request, handling it and writing a response to another thread (3) from the `handlerThreadPool`. Now, when our server blocks on slow database or client, or whatever, it actually blocks a thread from handler thread pool, while we still can accept connections in accepting connections thread pool.
```

Comment:
Tried to find code reference. Got lost in scary methods such as https://github.com/apache/tomcat70/blob/a545b443388186be3857acfbbd6e5a54f4b9dadf/java/org/apache/coyote/http11/AbstractHttp11Processor.java#L1267
I think I'll skip mentioning that it works like BIO from Tomcat since I'm not 100% sure how it works :D I may get back to it when I start coding NIO version (if ever 🤞)